packages/js/product-editor/src/blocks/variation-items/edit.tsx

```diff
@@ -19,11 +35,160 @@ export function Edit( {
 		isInSelectedTab?: boolean;
 	};
 } ) {
+	const noticeDimissed = useRef( false );
+	const { invalidateResolution } = useDispatch(
+		EXPERIMENTAL_PRODUCT_VARIATIONS_STORE_NAME
+	);
+	const productId = useEntityId( 'postType', 'product' );
 	const blockProps = useBlockProps();
+	const [ productStatus ] = useEntityProp< string >(
+		'postType',
+		'product',
+		'status'
+	);
+
+	const totalCountWithoutPriceRequestParams = useMemo(
+		() => ( {
+			product_id: productId,
+			order: 'asc',
+			orderby: 'menu_order',
+			has_price: false,
+		} ),
+		[ productId ]
+	);
+
+	const { totalCountWithoutPrice } = useSelect(
+		( select ) => {
+			const { getProductVariationsTotalCount } = select(
+				EXPERIMENTAL_PRODUCT_VARIATIONS_STORE_NAME
+			);
+
+			return {
+				totalCountWithoutPrice:
+					getProductVariationsTotalCount< number >(
+						totalCountWithoutPriceRequestParams
+					),
+			};
+		},
+		[ productId ]
+	);
+
+	const {
+		updateUserPreferences,
+		variable_items_without_price_notice_dismissed:
+			itemsWithoutPriceNoticeDismissed,
+	} = useUserPreferences();
+
+	const { ref: variationTableRef } = useValidation< Product >(
+		`variations`,
+		async function regularPriceValidator( defaultValue, additionalData ) {
+			if (
+				totalCountWithoutPrice > 0 &&
+				! noticeDimissed.current &&
+				productStatus !== 'publish' &&
+				additionalData?.status === 'publish'
+			) {
+				if ( itemsWithoutPriceNoticeDismissed !== 'yes' ) {
+					updateUserPreferences( {
+						variable_items_without_price_notice_dismissed: {
+							...( itemsWithoutPriceNoticeDismissed || {} ),
+							[ productId ]: 'no',
+						},
+					} );
+				}
+				return __(
+					'Set variation prices before adding this product.',
+					'woocommerce'
+				);
+			}
+		},
+		[ totalCountWithoutPrice ]
+	);
+
+	function onSetPrices(
```

Comment:
Nitpick: I would prefer having this directly in noticeActions prop. It would make it clearer to me that it's related to the VariationsTable component and you wouldn't even have to add type definitions since they would be inferred. But I understand if you want to leave it like this, it might be just personal preference.

Reply:
Hmm yeah, I would prefer to keep it as is, as `onSetPrices` is on the bigger side, which would make the props below a lot more complicated to read if I move the function inline.

Reply:
If it was smaller then yes, I agree inline would be nicer.